src/ds/address.h

```diff
@@ -0,0 +1,42 @@
+#pragma once
+#include <cstdint>
+
+namespace snmalloc
+{
+  /**
+   * The type used for an address.  Currently, all addresses are assumed to be
+   * provenance-carrying values and so it is possible to cast back from the
+   * result of arithmetic on an address_t.  Eventually, this will want to be
+   * separated into two types, one for raw addresses and one for addresses that
+   * can be cast back to pointers.
+   */
+  typedef uintptr_t address_t;
+
+  /**
+   * Perform pointer arithmetic and return the adjusted pointer.
+   */
+  template<typename T>
+  inline T* pointer_offset(T* base, size_t diff)
```

Comment:
size_t or ptrdiff_t?  I am happy with size_t.

Reply:
Callers pass `size_t` - I made it `ptrdiff_t` initially, but got lots of warnings about sign conversions.  I'd like to make it `ptrdiff_t` eventually, but it requires cleaning up all of the callers.